src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarHelpers.Unix.cs

```diff
@@ -0,0 +1,150 @@
+            => new SortedDictionary<string, UnixFileMode>(s_reverseStringComparer);
+
+        internal static void CreateDirectory(string fullPath, UnixFileMode? mode, bool overwriteMetadata, SortedDictionary<string, UnixFileMode>? pendingModes)
+        {
+            // Restrictive mask for creating the missing parent directories while extracting.
+            const UnixFileMode ExtractPermissions = UnixFileMode.UserRead | UnixFileMode.UserWrite | UnixFileMode.UserExecute;
+
+            Debug.Assert(pendingModes is not null);
+
+            if (Directory.Exists(fullPath))
+            {
+                // Apply permissions to an existing directory when we're overwriting metadata
+                // or the directory was created as a missing parent (stored in pendingModes).
+                if (mode.HasValue)
+                {
+                    bool hasExtractPermissions = (mode & ExtractPermissions) == ExtractPermissions;
+                    if (hasExtractPermissions)
+                    {
+                        bool removed = pendingModes.Remove(fullPath);
+                        if (overwriteMetadata || removed)
+                        {
+                            UnixFileMode umask = UMask;
+                            File.SetUnixFileMode(fullPath, mode.Value & ~umask);
+                        }
+                    }
+                    else if (overwriteMetadata || pendingModes.ContainsKey(fullPath))
+                    {
+                            pendingModes[fullPath] = mode.Value;
+                    }
+                }
+                return;
+            }
+
+            if (mode.HasValue)
+            {
+                // Ensure we have sufficient permissions to extract in the directory.
+                if ((mode & ExtractPermissions) != ExtractPermissions)
+                {
+                    pendingModes[fullPath] = mode.Value;
+                    mode = ExtractPermissions;
+                }
+            }
+            else
+            {
+                pendingModes.Add(fullPath, DefaultDirectoryMode);
+                mode = ExtractPermissions;
+            }
+
+            string parentDir = Path.GetDirectoryName(fullPath)!;
+            string rootDir = Path.GetPathRoot(parentDir)!;
+            bool hasMissingParents = false;
+            for (string dir = parentDir; dir != rootDir && !Directory.Exists(dir); dir = Path.GetDirectoryName(dir)!)
+            {
+                pendingModes.Add(dir, DefaultDirectoryMode);
+                hasMissingParents = true;
+            }
+
+            if (hasMissingParents)
+            {
+                Directory.CreateDirectory(parentDir, ExtractPermissions);
+            }
+
+            Directory.CreateDirectory(fullPath, mode.Value);
+        }
+
+        internal static void SetPendingModes(SortedDictionary<string, UnixFileMode>? pendingModes)
+        {
+            Debug.Assert(!OperatingSystem.IsWindows());
```

Comment:
This assert isn't needed, since the `TarHelpers.Unix.cs` file is only compiled in Unix.

Reply:
```suggestion
```